sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/ApproximateDistinct.java

```diff
@@ -0,0 +1,533 @@
+        throw new IllegalArgumentException("Coder is not deterministic ! " + e.getMessage(), e);
+      }
+      return new ApproximateDistinctFn<>(12, 0, coder);
+    }
+
+    /**
+     * Returns a new {@link ApproximateDistinctFn} combiner with a new precision p but
+     * the other parameters remain unchanged.
+     *
+     * @param p           the precision value for the normal representation
+     */
+    public ApproximateDistinctFn<InputT> withPrecision(int p) {
+      if (p < 4) {
+        throw new IllegalArgumentException(String.format("Expected : p >= 4. Actual : p = %s", p));
+      }
+      return new ApproximateDistinctFn<>(p, this.sp, this.inputCoder);
+    }
+
+    /**
+     * Returns a new {@link ApproximateDistinctFn} combiner with a sparse representation
+     * of precision sp but the other parameters remain unchanged.
+     *
+     * @param sp          the precision of HyperLogLog+' sparse representation
+     */
+    public ApproximateDistinctFn<InputT> withSparseRepresentation(int sp) {
+      if ((sp < this.p || sp > 32) && (sp != 0)) {
+          throw new IllegalArgumentException(String.format("Expected : p <= sp <= 32."
+                  + "Actual : p = %s, sp = %s", this.p, sp));
+      }
+      return new ApproximateDistinctFn<>(this.p, sp, this.inputCoder);
+    }
+
+    @Override
+    public HyperLogLogPlus createAccumulator() {
+      return new HyperLogLogPlus(p, sp);
+    }
+
+    @Override
+    public HyperLogLogPlus addInput(HyperLogLogPlus acc, InputT record) {
+      acc.offer(getBytes(record));
+      return acc;
+    }
+
+    private byte[] getBytes(InputT input) {
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      try {
+        inputCoder.encode(input, baos);
+        } catch (IOException e) {
+        throw new IllegalStateException(
+                "The input value cannot be encoded : " + e.getMessage(), e);
+        }
+      return baos.toByteArray();
+      }
+
+    /**
+     * Output the whole structure so it can be queried, reused or stored easily.
+     */
+    @Override
+    public HyperLogLogPlus extractOutput(HyperLogLogPlus accumulator) {
+      return accumulator;
+    }
+
+    @Override
+    public HyperLogLogPlus mergeAccumulators(Iterable<HyperLogLogPlus> accumulators) {
+      HyperLogLogPlus mergedAccum = createAccumulator();
+      for (HyperLogLogPlus accum : accumulators) {
+        try {
+          mergedAccum.addAll(accum);
+        } catch (CardinalityMergeException e) {
+          // Should never happen because only HyperLogLogPlus accumulators are instantiated.
+          throw new IllegalStateException("The accumulators cannot be merged : " + e.getMessage());
+        }
+      }
+      return mergedAccum;
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder
+              .add(DisplayData.item("inputCoder", inputCoder.getEncodedTypeDescriptor().toString())
```

Comment:
This item is not useful - the user probably knows what type of data they're applying this to.